File: tests-csharp/Pages/LoginPage.cs
using Microsoft.Playwright;
using Tests.CSharp.Pages;

namespace Tests.CSharp.Pages;

public class LoginPage : BasePage
{
    private readonly ILocator _usernameInput;
    private readonly ILocator _passwordInput;
    private readonly ILocator _loginButton;
    private readonly ILocator _errorMessage;

    public LoginPage(IPage page) : base(page)
    {
        _usernameInput = Page.Locator("[data-test='username']");
        _passwordInput = Page.Locator("[data-test='password']");
        _loginButton = Page.Locator("[data-test='login-button']");
        _errorMessage = Page.Locator("[data-test='error']");
    }

    public async Task LoginAsync(string username, string password)
    {
        await FillElementAsync(_usernameInput, username);
        await FillElementAsync(_passwordInput, password);
        await ClickElementAsync(_loginButton);
        await Page.WaitForLoadStateAsync(LoadState.NetworkIdle);
    }

    public async Task<bool> IsLoginButtonVisibleAsync()
    {
        return await IsElementVisibleAsync(_loginButton);
    }

    public async Task<string> GetErrorMessageAsync()
    {
        try
        {
            await WaitForElementAsync(_errorMessage, 5000);
            return await GetElementTextAsync(_errorMessage);
        }
        catch
        {
            return string.Empty;
        }
    }

    public async Task<bool> IsErrorMessageVisibleAsync()
    {
        return await IsElementVisibleAsync(_errorMessage);
    }

    public async Task ClearFormAsync()
    {
        await _usernameInput.ClearAsync();
        await _passwordInput.ClearAsync();
    }
}

File: tests-playwright/pages/login-page.ts
import { Page, Locator } from '@playwright/test';
import { BasePage } from './base-page';
import { User } from '../fixtures/test-data';

export class LoginPage extends BasePage {
  private readonly usernameInput: Locator;
  private readonly passwordInput: Locator;
  private readonly loginButton: Locator;
  private readonly errorMessage: Locator;

  constructor(page: Page) {
    super(page);
    this.usernameInput = page.locator('#user-name');
    this.passwordInput = page.locator('#password');
    this.loginButton = page.locator('#login-button');
    this.errorMessage = page.locator('[data-test="error"]');
  }

  async login(user: User): Promise<void> {
    await this.fillInput(this.usernameInput, user.username);
    await this.fillInput(this.passwordInput, user.password);
    await this.clickElement(this.loginButton);
  }

  async getErrorMessage(): Promise<string> {
    return await this.getElementText(this.errorMessage);
  }

  async isErrorMessageVisible(): Promise<boolean> {
    return await this.isElementVisible(this.errorMessage);
  }

  async isLoginButtonEnabled(): Promise<boolean> {
    return await this.loginButton.isEnabled();
  }

  async clearForm(): Promise<void> {
    await this.usernameInput.clear();
    await this.passwordInput.clear();
  }

  async getUsernameValue(): Promise<string> {
    return await this.usernameInput.inputValue();
  }

  async getPasswordValue(): Promise<string> {
    return await this.passwordInput.inputValue();
  }
}
